File: ios/InkFlowAI/Core/APIClient.swift
import Foundation

actor APIClient {
    static let shared = APIClient()

    private var accessToken: String? {
        KeychainHelper.load(account: "accessToken")
    }

    private var refreshToken: String? {
        KeychainHelper.load(account: "refreshToken")
    }

    func setTokens(access: String, refresh: String) {
        KeychainHelper.save(access, account: "accessToken")
        KeychainHelper.save(refresh, account: "refreshToken")
    }

    func clearTokens() {
        KeychainHelper.clearAuth()
    }

    // MARK: - Auth

    func login(email: String, password: String) async throws -> AuthResponse {
        try await post("/api/mobile/login", body: ["email": email, "password": password], auth: false)
    }

    func register(email: String, password: String, name: String?) async throws -> AuthResponse {
        var body: [String: String] = ["email": email, "password": password]
        if let name, !name.isEmpty { body["name"] = name }
        return try await post("/api/mobile/register", body: body, auth: false)
    }

    func refreshSession() async throws -> AuthResponse {
        guard let refresh = refreshToken else {
            throw APIError(message: "No refresh token", code: "NO_REFRESH")
        }
        return try await post(
            "/api/mobile/refresh",
            body: ["refreshToken": refresh],
            auth: false
        )
    }

    func logout() async {
        if let refresh = refreshToken {
            _ = try? await post(
                "/api/mobile/logout",
                body: ["refreshToken": refresh],
                auth: false
            ) as AuthResponse
        }
        clearTokens()
    }

    func fetchMe() async throws -> UserDTO {
        let res: MeResponse = try await get("/api/mobile/me")
        guard let user = res.user else {
            throw APIError(message: res.error ?? "Unauthorized", code: res.code)
        }
        return user
    }

    func fetchCredits() async throws -> CreditsResponse {
        try await get("/api/credits")
    }

    func fetchAppleProducts() async throws -> [AppleProductDTO] {
        let res: AppleProductsResponse = try await get("/api/apple/products")
        return res.products ?? []
    }

    func verifyAppleTransaction(
        transactionId: String,
        productId: String,
        signedTransaction: String? = nil
    ) async throws -> VerifyTransactionResponse {
        var body: [String: Any] = [
            "transactionId": transactionId,
            "productId": productId,
        ]
        if let signedTransaction {
            body["signedTransaction"] = signedTransaction
        }
        return try await post("/api/apple/verify-transaction", body: body)
    }

    func generateFinalInk(text: String, baseId: String = "poet") async throws -> GenerateResponse {
        try await post(
            "/api/generate",
            body: [
                "text": text,
                "baseId": baseId,
                "fluidity": 50,
                "rhythm": 50,
                "pressure": 50,
            ]
        )
    }

    func fetchSignatures() async throws -> [SavedSignatureDTO] {
        let res: SignaturesListResponse = try await get("/api/signatures")
        return res.signatures ?? []
    }

    func saveSignature(name: String, strokeData: StrokeDataDTO) async throws -> SaveSignatureResponse {
        let encoder = JSONEncoder()
        let strokeJSON = try JSONSerialization.jsonObject(with: encoder.encode(strokeData))
        return try await post(
            "/api/signatures",
            body: [
                "name": name,
                "strokeData": strokeJSON,
            ]
        )
    }

    func renameSignature(id: String, name: String) async throws {
        let _: OkResponse = try await patch(
            "/api/signatures/\(id)",
            body: ["name": name]
        )
    }

    func deleteSignature(id: String) async throws {
        let _: OkResponse = try await delete("/api/signatures/\(id)")
    }

    func analyzeRefine(imageBase64: String) async throws -> RefineResponse {
        try await post(
            "/api/refine",
            body: ["imageBase64": imageBase64]
        )
    }

    func signPDF(
        pdfBase64: String,
        signaturePngBase64: String,
        pageIndex: Int,
        x: Double,
        y: Double,
        width: Double,
        height: Double,
        fileName: String
    ) async throws -> SignPDFResponse {
        try await post(
            "/api/sign/pdf",
            body: [
                "pdfBase64": pdfBase64,
                "signaturePngBase64": signaturePngBase64,
                "pageIndex": pageIndex,
                "x": x,
                "y": y,
                "width": width,
                "height": height,
                "fileName": fileName,
                "sesAccepted": true,
            ]
        )
    }

    func deleteAccount() async throws {
        let _: OkResponse = try await delete("/api/account")
        clearTokens()
    }

    func requestPasswordReset(email: String) async throws {
        let _: OkResponse = try await post(
            "/api/auth/forgot-password",
            body: ["email": email],
            auth: false
        )
    }

    // MARK: - HTTP

    private func get<T: Decodable>(_ path: String) async throws -> T {
        try await request(path: path, method: "GET", body: nil as [String: String]?, retryOn401: true)
    }

    private func post<T: Decodable, B: Encodable>(
        _ path: String,
        body: B,
        auth: Bool = true
    ) async throws -> T {
        try await request(path: path, method: "POST", body: body, retryOn401: auth)
    }

    func post<T: Decodable>(
        _ path: String,
        body: [String: Any],
        auth: Bool = true
    ) async throws -> T {
        try await requestJSON(path: path, method: "POST", jsonBody: body, retryOn401: auth)
    }

    func patch<T: Decodable>(_ path: String, body: [String: Any]) async throws -> T {
        try await requestJSON(path: path, method: "PATCH", jsonBody: body, retryOn401: true)
    }

    func delete<T: Decodable>(_ path: String) async throws -> T {
        try await requestJSON(path: path, method: "DELETE", jsonBody: nil, retryOn401: true)
    }

    private func request<T: Decodable, B: Encodable>(
        path: String,
        method: String,
        body: B?,
        retryOn401: Bool
    ) async throws -> T {
        if let body {
            let data = try JSONEncoder().encode(body)
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
            return try await requestJSON(path: path, method: method, jsonBody: json, retryOn401: retryOn401)
        }
        return try await requestJSON(path: path, method: method, jsonBody: nil, retryOn401: retryOn401)
    }

    private func requestJSON<T: Decodable>(
        path: String,
        method: String,
        jsonBody: [String: Any]?,
        retryOn401: Bool,
        isRetry: Bool = false
    ) async throws -> T {
        var url = AppConfig.apiBaseURL
        url.append(path: path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")

        if let token = accessToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let jsonBody {
            req.httpBody = try JSONSerialization.data(withJSONObject: jsonBody)
        }

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw APIError(message: "Invalid response", code: nil)
        }

        if http.statusCode == 401, retryOn401, !isRetry {
            let refreshed = try await refreshSession()
            if refreshed.ok, let access = refreshed.accessToken, let refresh = refreshed.refreshToken {
                setTokens(access: access, refresh: refresh)
                return try await requestJSON(
                    path: path,
                    method: method,
                    jsonBody: jsonBody,
                    retryOn401: false,
                    isRetry: true
                )
            }
        }

        if http.statusCode >= 400 {
            if let err = try? JSONDecoder().decode(AuthResponse.self, from: data) {
                throw APIError(message: err.error ?? "Request failed", code: err.code)
            }
            throw APIError(message: "HTTP \(http.statusCode)", code: nil)
        }

        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: android/app/src/main/java/com/inkflow/ai/core/ApiClient.kt
package com.inkflow.ai.core

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.io.IOException
import java.lang.reflect.Type
import java.util.concurrent.TimeUnit

class ApiClient(private val tokenStore: TokenStore) {
    companion object {
        const val OFFLINE_CODE = "NETWORK_UNAVAILABLE"
        const val OFFLINE_MESSAGE =
            "No internet connection. Please connect to a network and try again."
    }

    private val gson = Gson()
    private val jsonMedia = "application/json; charset=utf-8".toMediaType()
    private val refreshMutex = Mutex()

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .addInterceptor(
            HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BASIC
            },
        )
        .build()

    fun setTokens(access: String, refresh: String) {
        tokenStore.accessToken = access
        tokenStore.refreshToken = refresh
    }

    fun clearTokens() {
        tokenStore.clear()
    }

    suspend fun login(email: String, password: String): AuthResponse =
        post("/api/mobile/login", mapOf("email" to email, "password" to password), auth = false)

    suspend fun register(email: String, password: String, name: String?): AuthResponse {
        val body = mutableMapOf<String, Any>("email" to email, "password" to password)
        if (!name.isNullOrBlank()) body["name"] = name
        return post("/api/mobile/register", body, auth = false)
    }

    suspend fun refreshSession(): AuthResponse {
        val refresh = tokenStore.refreshToken
            ?: throw ApiException("No refresh token", "NO_REFRESH")
        return post("/api/mobile/refresh", mapOf("refreshToken" to refresh), auth = false)
    }

    suspend fun logout() {
        val refresh = tokenStore.refreshToken
        if (refresh != null) {
            runCatching {
                post<AuthResponse>("/api/mobile/logout", mapOf("refreshToken" to refresh), auth = false)
            }
        }
        clearTokens()
    }

    suspend fun fetchMe(): UserDto {
        val res: MeResponse = get("/api/mobile/me")
        return res.user ?: throw ApiException(res.error ?: "Unauthorized", res.code)
    }

    suspend fun generateFinalInk(
        text: String,
        baseId: String = "poet",
        fluidity: Double = 50.0,
        rhythm: Double = 50.0,
        pressure: Double = 50.0,
        slant: Double = 0.0,
        size: Double = 1.0,
    ): GenerateResponse = post(
        "/api/generate",
        mapOf(
            "text" to text,
            "baseId" to baseId,
            "fluidity" to fluidity,
            "rhythm" to rhythm,
            "pressure" to pressure,
            "slant" to slant,
            "size" to size,
        ),
    )

    suspend fun fetchSignatures(): List<SavedSignatureDto> {
        val res: SignaturesListResponse = get("/api/signatures")
        return res.signatures.orEmpty()
    }

    suspend fun saveSignature(name: String, strokeData: StrokeDataDto): SaveSignatureResponse {
        val strokeJson = gson.toJsonTree(strokeData)
        return post(
            "/api/signatures",
            mapOf("name" to name, "strokeData" to strokeJson),
        )
    }

    /**
     * Saves straight from the studio settings — the server builds the stroke
     * payload and charges the 1-credit cloud-save fee. No paid generate needed.
     */
    suspend fun saveSignatureFromSettings(
        name: String,
        settings: SignatureSettingsDto,
        canvasWidth: Int,
        canvasHeight: Int,
    ): SaveSignatureResponse = post(
        "/api/signatures",
        mapOf(
            "name" to name,
            "settings" to gson.toJsonTree(settings),
            "canvasWidth" to canvasWidth,
            "canvasHeight" to canvasHeight,
        ),
    )

    /** Stores a photo-extracted transparent PNG in the cloud library (1 credit). */
    suspend fun saveCapturedSignature(
        name: String,
        capturedImage: String,
        canvasWidth: Int,
        canvasHeight: Int,
    ): SaveSignatureResponse = post(
        "/api/signatures",
        mapOf(
            "name" to name,
            "kind" to "captured",
            "capturedImage" to capturedImage,
            "canvasWidth" to canvasWidth,
            "canvasHeight" to canvasHeight,
        ),
    )

    suspend fun deleteSignature(id: String) {
        delete<OkResponse>("/api/signatures/$id")
    }

    suspend fun deleteAccount() {
        delete<OkResponse>("/api/account")
        clearTokens()
    }

    suspend fun fetchGoogleProducts(): List<GoogleProductDto> {
        val res: GoogleProductsResponse = get("/api/google/products")
        return res.products.orEmpty()
    }

    suspend fun verifyGooglePurchase(
        productId: String,
        purchaseToken: String,
        productType: String,
    ): VerifyGoogleResponse = post(
        "/api/google/verify-purchase",
        mapOf(
            "productId" to productId,
            "purchaseToken" to purchaseToken,
            "productType" to productType,
        ),
    )

    suspend fun forgotPassword(email: String) {
        post<OkResponse>("/api/auth/forgot-password", mapOf("email" to email), auth = false)
    }

    suspend fun signPdf(
        pdfBase64: String,
        signaturePngBase64: String,
        pageIndex: Int,
        x: Double,
        y: Double,
        width: Double,
        height: Double,
        fileName: String,
    ): SignPdfResponse = post(
        "/api/sign/pdf",
        mapOf(
            "pdfBase64" to pdfBase64,
            "signaturePngBase64" to signaturePngBase64,
            "pageIndex" to pageIndex,
            "x" to x,
            "y" to y,
            "width" to width,
            "height" to height,
            "fileName" to fileName,
            "sesAccepted" to true,
        ),
    )

    /** Keeps a finished PDF in the cloud library. Free — signing already charged. */
    suspend fun uploadDocument(fileName: String, pdfBase64: String): DocumentResponse =
        post(
            "/api/documents",
            mapOf("fileName" to fileName, "pdfBase64" to pdfBase64),
        )

    suspend fun fetchDocuments(): List<CloudDocumentDto> {
        val res: DocumentsListResponse = get("/api/documents")
        return res.documents.orEmpty()
    }

    suspend fun fetchDocument(id: String): CloudDocumentDto {
        val res: DocumentResponse = get("/api/documents/$id")
        return res.document
            ?: throw ApiException(res.error ?: "Document not found.", res.code)
    }

    suspend fun deleteDocument(id: String) {
        delete<OkResponse>("/api/documents/$id")
    }

    private suspend inline fun <reified T> get(path: String): T =
        request(path, "GET", null, retryOn401 = true)

    private suspend inline fun <reified T> post(
        path: String,
        body: Map<String, Any?>,
        auth: Boolean = true,
    ): T = request(path, "POST", body, retryOn401 = auth)

    private suspend inline fun <reified T> delete(path: String): T =
        request(path, "DELETE", null, retryOn401 = true)

    private suspend inline fun <reified T> request(
        path: String,
        method: String,
        body: Map<String, Any?>?,
        retryOn401: Boolean,
    ): T {
        val type = object : TypeToken<T>() {}.type
        return requestTyped(path, method, body, retryOn401, type)
    }

    private suspend fun <T> requestTyped(
        path: String,
        method: String,
        body: Map<String, Any?>?,
        retryOn401: Boolean,
        type: Type,
        isRetry: Boolean = false,
    ): T = withContext(Dispatchers.IO) {
        val url = AppConfig.API_BASE_URL.trimEnd('/') + "/" + path.trimStart('/')
        val builder = Request.Builder().url(url)
        tokenStore.accessToken?.let {
            builder.header("Authorization", "Bearer $it")
        }
        builder.header("Accept", "application/json")
        if (body != null) {
            val json = gson.toJson(body)
            builder.method(method, json.toRequestBody(jsonMedia))
            builder.header("Content-Type", "application/json")
        } else {
            builder.method(
                method,
                if (method == "GET" || method == "DELETE") null else "".toRequestBody(jsonMedia),
            )
        }

        // A dropped connection reads as "Unable to resolve host …" out of OkHttp;
        // translate it into something a user can act on.
        val response = try {
            client.newCall(builder.build()).execute()
        } catch (_: IOException) {
            throw ApiException(OFFLINE_MESSAGE, OFFLINE_CODE)
        }
        val bytes = try {
            response.body?.string().orEmpty()
        } catch (_: IOException) {
            throw ApiException(OFFLINE_MESSAGE, OFFLINE_CODE)
        }

        if (response.code == 401 && retryOn401 && !isRetry) {
            refreshMutex.withLock {
                val refreshed = refreshSession()
                if (refreshed.ok && refreshed.accessToken != null && refreshed.refreshToken != null) {
                    setTokens(refreshed.accessToken, refreshed.refreshToken)
                    return@withContext requestTyped(
                        path,
                        method,
                        body,
                        retryOn401 = false,
                        type = type,
                        isRetry = true,
                    )
                }
            }
        }

        if (response.code >= 400) {
            val err = runCatching { gson.fromJson(bytes, AuthResponse::class.java) }.getOrNull()
            throw ApiException(
                err?.error ?: "HTTP ${response.code}",
                err?.code,
            )
        }

        @Suppress("UNCHECKED_CAST")
        (gson.fromJson<T>(bytes, type) ?: throw ApiException("Empty response"))
    }
}
